File: R/Tutorial 1.R
### Preliminaries ###

getwd()                                                                     # Check the working directory 
setwd("")                                                                   # Set the working directory
rm(list=ls())                                                               # Clear all 


### Installing and Loading Libraries 

# Loading the tidyverse package for data manipulation
library(tidyverse)
# Loading the rsample package for data splitting
library(rsample)
# Loading the sandwich package for robust standard errors
library(sandwich)
# Loading the lmtest package for hypothesis testing 
library(lmtest)

### Loading, Exporting and Importing Data ###            

# Viewing all built-in R data sets, and datasets that are included with packages        
data()
# Loading the mtcars data that comes with R
data(mtcars)
# Viewing the data 
View(mtcars)
# Investigating the structure of the data 
str(mtcars)
# Viewing the data documentation
?mtcars
# Viewing the help file for the str() command 
?str
# Gives us a top level overview of the data set 
glimpse(mtcars)
# Prints the first 5 observations
head(mtcars,5)
# Creating a dataframe object with the mtcars data called "df"
df <- as.data.frame(mtcars)
# Creating a dataframe object with the mtcars data called "df"
df = as.data.frame(mtcars)
# exporting a dataframe object as a .csv file 
write.csv(df,"mtcars_export.csv", row.names = TRUE)   
# importing a .csv file as a dataframe object          
df_import <- data.frame(read.csv("mtcars_export.csv"))
# Atomic vector, R's most basic data structure, More on this later.
c(1,2)
         
### Data Manipulation          

## Basic Functions

# Looking at column names
names(df)
# Check if any observations are N/A's
is.na(df)
# Investigating the number of N/A's in each column 
colSums(is.na(df))
# Looking at the dimensions of an object
dim(df)

## Slicing

# Selecting a specific column using the column name 
df$cyl
# Viewing a Column Slice 
mtcars[1]
# Looking at a slice with 2 columns
mtcars[,c("mpg", "cyl")]
# Viewing a Row Slice
mtcars[c("Mazda RX4"),]
# viewing a specific column using the column index 
mtcars[[1]]
# Viewing a specific column using the column name
mtcars[["cyl"]]
         
## More Advanced Manipulation

# Keep rows that match a column condition
filter(df, am == 1)
# Average mpg by variable am
df %>% group_by(am) %>% summarise(mean(mpg))
# Converting mpg to gpm and adding it as a new column
df <- mutate(df, gpm = 1/mpg)
head(df)
# Renaming variables, the pipe operator passes the df object as the first argument in the rename() function
df <- df %>% rename(horsepower = hp)
# Selecting a few columns to keep them in the data frame
df_selected <- df %>% select(mpg, horsepower)
# Extracting a column and creating a vector
hp  <- as.vector(df$horsepower)
# Creating dummy variables 
btb <- ifelse(horsepower>200, 1, 0)
# Summarizing the dummy variable 
summary(btb)
# Adding this object to the data frame
df_btb <- cbind(btb, df)
# Removing variables
df_sel <- df_btb %>% select(-mpg, -cyl)

### Data visualization###

## Base R
# Histogram
hist(df$mpg,xlab="miles per gallon",main="Histogram of mpg")
# Scatter plot
plot(x=df$wt,y=df$mpg,xlab="weight",ylab="miles per gallon")

## ggplot2 package
# First we provide the data, and then define the x-axis variable
ggplot(df, aes(x = mpg)) +
  # Command to plot a histogram
  geom_histogram()
# Providing the data, and defining the x - and y - axis variables 
ggplot(df, aes(x = wt, y = mpg, color = am)) +
  # Command for a scatter plot
  geom_point()
# Providing the data, and defining the x - and y - axis variables
ggplot(df, aes(x = wt, y = mpg)) +
  # Command for a scatter plot
  geom_point() +
  # Smoothing command, specifying that we want to use the method "lm" i.e., regression line
  geom_smooth(method = lm) +
  # Defining the axis labels
  labs(x = "weight", y = "miles per gallon")

### Data Splitting ###

# Using rsample 
# Defining the split proportion 
df_split <- initial_split(df, prop=.7)
# Extracting the training sample
df_train <- training(df_split)
# Extracting the testing sample
df_test  <- testing(df_split)
                    
# Using baseR                                     
# Defining the fraction of data to be used as the training sample
t_f <- 0.70
# Saving the number of observations as an object called "n"
n <- dim(df)[1]
# The sample.int function takes a sample of the specified size 
train_idx <- sample.int(n,
                        # Specifying that we want to sample without replacement
                        replace=FALSE,
                        # Specifying the number of observations to choose 
                        size=floor(n*t_f))
# Subsetting and saving the training sample
df_train_2 <- df[train_idx,]
# Subsetting and saving the testing sample
df_test_2  <- df[-train_idx,]

### OLS Regressions ### 
# Specifying the linear model, and saving the regression model
ols_1 <- lm(mpg ~ cyl + horsepower + am, data=df)
# Looking at the regression Summary
summary(ols_1)
# Extracting Hatvalues, play around with this, read the documentation available online
hatvalues(ols_1)
# Getting Robust Standard Errors
coeftest(ols_1, vcov = vcovHC(ols_1, type = 'HC3'))

File: Python/Tutorial 1.py
### Preliminaries ###
# Loading os library to interact with system
import os
# Check the working directory
os.getcwd()
# Set the working directory
os.chdir("")

### Loading Libraries ###
# Loading the pandas library for data manipulation
import pandas as pd
# Loading the numpy package for matrices and many useful mathematical functions
import numpy as np
# Loading the matplotlib package for plotting
import matplotlib.pyplot as plt
# Loading the statsmodels package for statistical models
import statsmodels.api as sm
# Loading the statsmodels formula package for R-like formula functionality
import statsmodels.formula.api as smf
# Loading data splitting function from the sklearn package
from sklearn.model_selection import train_test_split

### Loading, Exporting, and Importing Data ###

# Importing a .csv file as a pandas dataframe
mtcars = pd.read_csv("../Datasets/mtcars.csv",index_col=0)
# Getting documentation for a function
help(pd.read_csv)
# Prints the first 5 observations
mtcars.head()
# Note that we did not type something like head(mtcars)
# This is because the head() function is associated with the pandas DataFrame "class".
# Getting documentation for head function
help(pd.DataFrame.head)
# Note that we had to add the function head()'s associated class using pd.DataFrame
# You can also just google a function's documentation quite easily.

# Saving a pandas dataframe as a .csv file
mtcars.to_csv("../Datasets/mtcars.csv")

### Data Manipulation ###

## Basic Functions ##
# Looking at column names
mtcars.columns
# Check if any observations are N/A's
mtcars.isnull()
# Number of N/A's in each column
mtcars.isnull().sum(axis=0)
# Dimension of dataframe
mtcars.shape

## Slicing ##
# Selecting a column using its name or 'index'
mtcars.cyl
mtcars['cyl']
# Selecting a column using its integer index using iloc
mtcars.iloc[:,1]
# The : before the comma indicates we want all rows.
# Selecting a column using its name with loc
mtcars.loc[:,'cyl']
# Selecting multiple columns
mtcars.loc[:,['mpg','cyl']]
# Selecting a row by its index
mtcars.loc['Mazda RX4']
# Note that in this case we did not need to specify any column indices, although we can.
mtcars.loc['Mazda RX4',:]

## More Advanced Manipulation

# Keep rows that match a column condition
mtcars.loc[mtcars.am==1]
# Thus, .loc can be used by specifying indices like in the slicing examples above,
# or by specifying a condition.

# Average mpg by the am variable
mtcars.groupby('am')['mpg'].mean()
# We grouped by the am variable, selected the mpg column, and then calculated mean.

# Coverting mpg to gpm and adding it as a new column
mtcars['gpm'] = 1/mtcars['mpg']
mtcars.head()
# Renaming variables
mtcars = mtcars.rename(columns={'hp':'horsepower'})
# Creating dummy variables
btb = (mtcars['horsepower']>200).astype(int)
# Summarizing the dummary variable
btb.describe()
# Adding it to the data frame
mtcars['btb'] = btb
# Removing variables
mtcars_sel = mtcars.drop(columns=['mpg','cyl'])

### Data visualization ###

## The most common data visualization package is matplotlib

# Creating a histogram
mtcars['mpg'].hist(edgecolor='black',grid=False)
# Adding axis labels
plt.xlabel('miles per gallon')
plt.title('Histogram of mpg')
plt.show()

# We can use .hist on the dataframe itself since pandas dataframes can
# call matplotlib functions. We can also use matplotlib directly.
plt.hist(mtcars['mpg'],edgecolor='black')
plt.xlabel('miles per gallon')
plt.title('Histogram of mpg')
plt.show()
# Note that the default arguments of .hist and matplotlib's hist are different.
# We did not need to specify grid=False with matplotlib.

# Creating a scatterplot
plt.scatter(x='wt',y='mpg',c='am',data=mtcars)
plt.colorbar(label='am')
plt.xlabel('Weight')
plt.ylabel('Miles per Gallon')
plt.title('mpg on weight')
plt.show()
# Note. colorbar is not a great way to show a legend for the categories.
# Making a legend for discrete categories is a bit more complicated with
# matplotlib. Seaborn is another great plotting package built on top of 
# matplotlib which simplifies the code greatly. More on this later.

### Data Splitting ###
# We will need the sklearn package for this
mtcars_train, mtcars_test = train_test_split(mtcars,test_size=0.3)

### OLS Regressions ###

# Fit the model
model = smf.ols('mpg ~ cyl + horsepower + am',data=mtcars).fit()
# Summary of regression results
print(model.summary())
# Extracting predicted values
model.predict(mtcars)
# Getting robust standard errors
model_hc3 = model.get_robustcov_results(cov_type='HC3')
print(model_hc3.summary())